beetsplug/rewrite.py

```diff
@@ -34,9 +34,7 @@ def rewriter(field, rules):
     def fieldfunc(item):
         value = item._values_fixed[field]
         for pattern, replacement in rules:
-            if pattern.match(value.lower()):
-                # Rewrite activated.
-                return replacement
+            return re.sub(pattern, replacement, value)
```

Comment:
Alright, here's an idea:

Generally we have to distinguish four cases:
- Search for entire static string, entirely replace it with a static string (`"The Jimi Hendrix Experience" -> "Jimi Hendrix"`)
- Search for regex, entirely replace it with a static string (`.*jimi hendrix.* -> "Jimi Hendrix"`)
- Search for regex, only replace substrings (`(.*)\(([0-9]{4})\) -> \1[\2]`)
- ~~Search for entire static string, only replace substrings~~ This is already covered by the first case, since the only possible substring would be the search string itself.

Since static search strings are technically regexes as well, the first two cases can be combined. We therefore need to distinguish between:
- Search for regex, entirely replace it with a static string (current behavior)
- Search for regex, replace substrings (addition)

So, how can we distinguish between a static replacement string and a dynamic replacement string?
At first, I thought we could test if the replacement string is a regex, but that seems to be impossible.
However, it might be a lot simpler: The only reason you would want to use a regex as a replacement pattern is that you want to modify existing information. This information needs to be back-referenced, which is done with the special strings `'\{match-position}'`, e.g. `\1`, `\2`, … etc. So I think it would be sufficient enough to check the replacement string for occurrences of this special string.
This can be done with …another regex, such as `'\\([0-9]+)'`.
- If this does not match the replacement pattern, we know that we want to replace the entire string, invoke the current behavior.
- If it does match, we replace substrings by running the new code.

This approach would also solve the case-sensitivity issue, since it wouldn't affect old rewrite rules. Moreover, everybody knows that regexes are case-sensitive, so new rules employing the back-referencing feature will be designed accordingly.

So where's the catch? Let me know what you think :)